File: src/main/java/com/uiptv/util/SQLiteTableSync.java
package com.uiptv.util;

import com.uiptv.db.DatabasePatchesUtils;
import com.uiptv.db.DatabaseUtils;

import java.io.File;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public final class SQLiteTableSync {
    private static final String SQLITE_PREFIX = "jdbc:sqlite:";

    private SQLiteTableSync() {
    }

    public static void ensureDatabaseReady(String dbPath) throws SQLException {
        try {
            File file = new File(dbPath);
            File parent = file.getParentFile();
            if (parent != null && !parent.exists() && !parent.mkdirs() && !parent.exists()) {
                throw new IOException("Unable to create parent directory for " + dbPath);
            }
            if (!file.exists() && !file.createNewFile() && !file.exists()) {
                throw new IOException("Unable to create database file " + dbPath);
            }
        } catch (IOException e) {
            throw new SQLException("Unable to prepare database file " + dbPath, e);
        }

        try (Connection conn = DriverManager.getConnection(SQLITE_PREFIX + dbPath)) {
            DatabasePatchesUtils.applyPatches(conn);
        }
    }

    public static int syncTables(String sourceDBPath, String targetDBPath, DatabaseUtils.DbTable table) throws SQLException {
        String tableName = DatabaseUtils.validatedTableName(table);
        try (
                Connection sourceConn = DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath);
                Connection targetConn = DriverManager.getConnection(SQLITE_PREFIX + targetDBPath)
        ) {
            int syncedRows = syncTable(sourceConn, targetConn, tableName);
            AppLog.addInfoLog(SQLiteTableSync.class, "Table '" + tableName + "' synced from source to target.");
            return syncedRows;
        }
    }

    public static int replaceTable(String sourceDBPath, String targetDBPath, DatabaseUtils.DbTable table) throws SQLException {
        String tableName = DatabaseUtils.validatedTableName(table);
        try (
                Connection sourceConn = DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath);
                Connection targetConn = DriverManager.getConnection(SQLITE_PREFIX + targetDBPath)
        ) {
            int syncedRows = replaceTable(sourceConn, targetConn, tableName);
            AppLog.addInfoLog(SQLiteTableSync.class, "Table '" + tableName + "' replaced from source to target.");
            return syncedRows;
        }
    }

    public static boolean syncConfiguration(String sourceDBPath, String targetDBPath, boolean includeExternalPlayerPaths) throws SQLException {
        try (
                Connection sourceConn = DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath);
                Connection targetConn = DriverManager.getConnection(SQLITE_PREFIX + targetDBPath)
        ) {
            boolean synced = syncConfiguration(sourceConn, targetConn, includeExternalPlayerPaths);
            AppLog.addInfoLog(SQLiteTableSync.class, "Configuration synced from source to target.");
            return synced;
        }
    }

    public static int syncPublishedM3uSelections(String sourceDBPath, String targetDBPath) throws SQLException {
        try (
                Connection sourceConn = DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath);
                Connection targetConn = DriverManager.getConnection(SQLITE_PREFIX + targetDBPath)
        ) {
            int synced = syncPublishedM3uSelections(sourceConn, targetConn);
            AppLog.addInfoLog(SQLiteTableSync.class, "PublishedM3uSelection synced from source to target with account remapping.");
            return synced;
        }
    }

    private static int syncTable(Connection sourceConn, Connection targetConn, String tableName) throws SQLException {
        List<String> targetColumns = getTableColumns(targetConn, tableName);
        Set<String> targetColumnSet = new LinkedHashSet<>(targetColumns);
        List<String> sourceColumns = getTableColumns(sourceConn, tableName);
        List<String> commonColumns = sourceColumns.stream()
                .filter(targetColumnSet::contains)
                .toList();

        if (commonColumns.isEmpty()) {
            throw new SQLException("No common columns found for table " + tableName);
        }

        String columnList = String.join(", ", commonColumns);
        String placeholders = commonColumns.stream().map(column -> "?").collect(Collectors.joining(", "));
        String selectSql = "SELECT " + columnList + " FROM " + tableName;
        String insertSql = "INSERT OR REPLACE INTO " + tableName + " (" + columnList + ") VALUES (" + placeholders + ")";

        try (
                Statement sourceStmt = sourceConn.createStatement();
                ResultSet sourceResult = sourceStmt.executeQuery(selectSql);
                PreparedStatement targetStatement = targetConn.prepareStatement(insertSql)
        ) {
            int columnCount = commonColumns.size();
            int syncedRows = 0;
            while (sourceResult.next()) {
                for (int i = 1; i <= columnCount; i++) {
                    targetStatement.setObject(i, sourceResult.getObject(i));
                }
                targetStatement.addBatch();
                syncedRows++;
            }
            targetStatement.executeBatch();
            return syncedRows;
        }
    }

    private static int replaceTable(Connection sourceConn, Connection targetConn, String tableName) throws SQLException {
        List<String> targetColumns = getTableColumns(targetConn, tableName);
        Set<String> targetColumnSet = new LinkedHashSet<>(targetColumns);
        List<String> sourceColumns = getTableColumns(sourceConn, tableName);
        List<String> commonColumns = sourceColumns.stream()
                .filter(targetColumnSet::contains)
                .toList();

        if (commonColumns.isEmpty()) {
            throw new SQLException("No common columns found for table " + tableName);
        }

        String columnList = String.join(", ", commonColumns);
        String placeholders = commonColumns.stream().map(column -> "?").collect(Collectors.joining(", "));
        String selectSql = "SELECT " + columnList + " FROM " + tableName;
        String deleteSql = "DELETE FROM " + tableName;
        String insertSql = "INSERT INTO " + tableName + " (" + columnList + ") VALUES (" + placeholders + ")";

        boolean originalAutoCommit = targetConn.getAutoCommit();
        try (
                Statement sourceStmt = sourceConn.createStatement();
                ResultSet sourceResult = sourceStmt.executeQuery(selectSql);
                Statement deleteStatement = targetConn.createStatement();
                PreparedStatement targetStatement = targetConn.prepareStatement(insertSql)
        ) {
            targetConn.setAutoCommit(false);
            deleteStatement.executeUpdate(deleteSql);

            int columnCount = commonColumns.size();
            int syncedRows = 0;
            while (sourceResult.next()) {
                for (int i = 1; i <= columnCount; i++) {
                    targetStatement.setObject(i, sourceResult.getObject(i));
                }
                targetStatement.addBatch();
                syncedRows++;
            }
            targetStatement.executeBatch();
            targetConn.commit();
            return syncedRows;
        } catch (SQLException e) {
            targetConn.rollback();
            throw e;
        } finally {
            targetConn.setAutoCommit(originalAutoCommit);
        }
    }

    private static boolean syncConfiguration(Connection sourceConn, Connection targetConn, boolean includeExternalPlayerPaths) throws SQLException {
        List<String> sourceColumns = getTableColumns(sourceConn, DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName());
        List<String> targetColumns = getTableColumns(targetConn, DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName());
        List<String> commonColumns = sourceColumns.stream()
                .filter(new LinkedHashSet<>(targetColumns)::contains)
                .toList();

        if (commonColumns.isEmpty()) {
            return false;
        }

        ConfigurationRow sourceRow = readFirstConfigurationRow(sourceConn, commonColumns);
        if (sourceRow == null) {
            return false;
        }
        ConfigurationRow targetRow = readFirstConfigurationRow(targetConn, commonColumns);
        if (!includeExternalPlayerPaths && targetRow != null) {
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath1");
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath2");
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath3");
            sourceRow.copyColumnIfPresentFrom(targetRow, "defaultPlayerPath");
        }

        upsertFirstConfigurationRow(targetConn, sourceRow, targetRow != null ? targetRow.id : null);
        return true;
    }

    private static int syncPublishedM3uSelections(Connection sourceConn, Connection targetConn) throws SQLException {
        String selectionTable = DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.getTableName();
        String accountTable = DatabaseUtils.DbTable.ACCOUNT_TABLE.getTableName();
        boolean originalAutoCommit = targetConn.getAutoCommit();
        int syncedRows = 0;

        try (
                PreparedStatement sourceSelections = sourceConn.prepareStatement(
                        "SELECT accountId FROM " + selectionTable + " ORDER BY id");
                ResultSet sourceRows = sourceSelections.executeQuery();
                PreparedStatement sourceAccountName = sourceConn.prepareStatement(
                        "SELECT accountName FROM " + accountTable + " WHERE id = ?");
                PreparedStatement targetAccountId = targetConn.prepareStatement(
                        "SELECT id FROM " + accountTable + " WHERE accountName = ?");
                Statement deleteTargetSelections = targetConn.createStatement();
                PreparedStatement insertTargetSelection = targetConn.prepareStatement(
                        "INSERT INTO " + selectionTable + " (accountId) VALUES (?)")
        ) {
            targetConn.setAutoCommit(false);
            deleteTargetSelections.executeUpdate("DELETE FROM " + selectionTable);

            while (sourceRows.next()) {
                String sourceAccountId = sourceRows.getString("accountId");
                if (sourceAccountId == null || sourceAccountId.isBlank()) {
                    continue;
                }

                sourceAccountName.setString(1, sourceAccountId);
                try (ResultSet sourceAccount = sourceAccountName.executeQuery()) {
                    if (!sourceAccount.next()) {
                        continue;
                    }

                    String accountName = sourceAccount.getString("accountName");
                    if (accountName == null || accountName.isBlank()) {
                        continue;
                    }

                    targetAccountId.setString(1, accountName);
                    try (ResultSet targetAccount = targetAccountId.executeQuery()) {
                        if (!targetAccount.next()) {
                            continue;
                        }
                        insertTargetSelection.setString(1, targetAccount.getString("id"));
                        insertTargetSelection.addBatch();
                        syncedRows++;
                    }
                }
            }

            if (syncedRows > 0) {
                insertTargetSelection.executeBatch();
            }
            targetConn.commit();
            return syncedRows;
        } catch (SQLException e) {
            targetConn.rollback();
            throw e;
        } finally {
            targetConn.setAutoCommit(originalAutoCommit);
        }
    }

    private static ConfigurationRow readFirstConfigurationRow(Connection conn, List<String> columns) throws SQLException {
        String columnList = columns.stream()
                .filter(column -> !"id".equalsIgnoreCase(column))
                .collect(Collectors.joining(", "));
        if (columnList.isBlank()) {
            return null;
        }

        String sql = "SELECT id, " + columnList + " FROM " + DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName() + " ORDER BY id LIMIT 1";
        try (Statement stmt = conn.createStatement(); ResultSet rs = stmt.executeQuery(sql)) {
            if (!rs.next()) {
                return null;
            }
            ConfigurationRow row = new ConfigurationRow(rs.getObject("id"));
            for (String column : columns) {
                if ("id".equalsIgnoreCase(column)) {
                    continue;
                }
                row.put(column, rs.getObject(column));
            }
            return row;
        }
    }

    private static void upsertFirstConfigurationRow(Connection conn, ConfigurationRow row, Object targetId) throws SQLException {
        List<String> columns = new ArrayList<>(row.columns());
        if (columns.isEmpty()) {
            return;
        }

        if (targetId == null) {
            String columnList = String.join(", ", columns);
            String placeholders = columns.stream().map(column -> "?").collect(Collectors.joining(", "));
            String sql = "INSERT INTO " + DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName()
                    + " (" + columnList + ") VALUES (" + placeholders + ")";
            try (PreparedStatement statement = conn.prepareStatement(sql)) {
                bindColumns(statement, columns, row);
                statement.executeUpdate();
            }
            return;
        }

        String assignments = columns.stream().map(column -> column + "=?").collect(Collectors.joining(", "));
        String sql = "UPDATE " + DatabaseUtils.DbTable.CONFIGURATION_TABLE.getTableName()
                + " SET " + assignments + " WHERE id = ?";
        try (PreparedStatement statement = conn.prepareStatement(sql)) {
            bindColumns(statement, columns, row);
            statement.setObject(columns.size() + 1, targetId);
            statement.executeUpdate();
        }
    }

    private static void bindColumns(PreparedStatement statement, List<String> columns, ConfigurationRow row) throws SQLException {
        for (int i = 0; i < columns.size(); i++) {
            statement.setObject(i + 1, row.get(columns.get(i)));
        }
    }

    private static List<String> getTableColumns(Connection conn, String tableName) throws SQLException {
        List<String> columns = new ArrayList<>();
        try (Statement stmt = conn.createStatement(); ResultSet rs = stmt.executeQuery("PRAGMA table_info(" + tableName + ")")) {
            while (rs.next()) {
                columns.add(rs.getString("name"));
            }
        }
        return columns;
    }

    private static final class ConfigurationRow {
        private final Object id;
        private final java.util.LinkedHashMap<String, Object> values = new java.util.LinkedHashMap<>();

        private ConfigurationRow(Object id) {
            this.id = id;
        }

        void put(String column, Object value) {
            values.put(column, value);
        }

        Object get(String column) {
            return values.get(column);
        }

        Set<String> columns() {
            return values.keySet();
        }

        void copyColumnIfPresentFrom(ConfigurationRow source, String column) {
            if (source != null && values.containsKey(column) && source.values.containsKey(column)) {
                values.put(column, source.values.get(column));
            }
        }
    }
}

File: server/src/main/kotlin/com/uiptv/util/SQLiteTableSync.kt
package com.uiptv.util

import com.uiptv.db.DatabasePatchesUtils
import com.uiptv.db.DatabaseUtils
import java.io.File
import java.io.IOException
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Statement
import java.util.stream.Collectors

object SQLiteTableSync {
    private const val SQLITE_PREFIX = "jdbc:sqlite:"
    private const val SQL_SELECT = "SELECT "
    private const val SQL_FROM = " FROM "
    private const val SQL_INSERT_INTO = "INSERT INTO "
    private const val SQL_VALUES = ") VALUES ("
    private const val ACCOUNT_ID = "accountId"
    private val CONFIGURATION_SYNC_EXCLUDED_COLUMNS = setOf("filterLockHash")

    @JvmStatic
    @Throws(SQLException::class)
    fun ensureDatabaseReady(dbPath: String) {
        try {
            val file = File(dbPath)
            val parent = file.parentFile
            if (parent != null && !parent.exists() && !parent.mkdirs() && !parent.exists()) {
                throw IOException("Unable to create parent directory for $dbPath")
            }
            if (!file.exists() && !file.createNewFile() && !file.exists()) {
                throw IOException("Unable to create database file $dbPath")
            }
        } catch (e: IOException) {
            throw SQLException("Unable to prepare database file $dbPath", e)
        }

        DriverManager.getConnection(SQLITE_PREFIX + dbPath).use { conn ->
            DatabasePatchesUtils.applyPatches(conn)
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncTables(sourceDBPath: String, targetDBPath: String, table: DatabaseUtils.DbTable): Int {
        val tableName = DatabaseUtils.validatedTableName(table)
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val syncedRows = syncTable(sourceConn, targetConn, tableName)
                AppLog.addInfoLog(SQLiteTableSync::class.java, "Table '$tableName' synced from source to target.")
                return syncedRows
            }
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun replaceTable(sourceDBPath: String, targetDBPath: String, table: DatabaseUtils.DbTable): Int {
        val tableName = DatabaseUtils.validatedTableName(table)
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val syncedRows = replaceTable(sourceConn, targetConn, tableName)
                AppLog.addInfoLog(SQLiteTableSync::class.java, "Table '$tableName' replaced from source to target.")
                return syncedRows
            }
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncConfiguration(sourceDBPath: String, targetDBPath: String, includeExternalPlayerPaths: Boolean): Boolean {
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val synced = syncConfiguration(sourceConn, targetConn, includeExternalPlayerPaths)
                AppLog.addInfoLog(SQLiteTableSync::class.java, "Configuration synced from source to target.")
                return synced
            }
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncPublishedM3uSelections(sourceDBPath: String, targetDBPath: String): Int {
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val synced = syncPublishedM3uSelections(sourceConn, targetConn)
                AppLog.addInfoLog(SQLiteTableSync::class.java, "PublishedM3uSelection synced from source to target with account remapping.")
                return synced
            }
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncPublishedM3uCategorySelections(sourceDBPath: String, targetDBPath: String): Int {
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val synced = syncSelectionTableWithAccountRemap(
                    sourceConn,
                    targetConn,
                    DatabaseUtils.DbTable.PUBLISHED_M3U_CATEGORY_SELECTION_TABLE.tableName,
                    listOf("categoryName", "selected")
                )
                AppLog.addInfoLog(SQLiteTableSync::class.java, "PublishedM3uCategorySelection synced from source to target with account remapping.")
                return synced
            }
        }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun syncPublishedM3uChannelSelections(sourceDBPath: String, targetDBPath: String): Int {
        DriverManager.getConnection(SQLITE_PREFIX + sourceDBPath).use { sourceConn ->
            DriverManager.getConnection(SQLITE_PREFIX + targetDBPath).use { targetConn ->
                val synced = syncSelectionTableWithAccountRemap(
                    sourceConn,
                    targetConn,
                    DatabaseUtils.DbTable.PUBLISHED_M3U_CHANNEL_SELECTION_TABLE.tableName,
                    listOf("categoryName", "channelId", "selected")
                )
                AppLog.addInfoLog(SQLiteTableSync::class.java, "PublishedM3uChannelSelection synced from source to target with account remapping.")
                return synced
            }
        }
    }

    @Throws(SQLException::class)
    private fun syncTable(sourceConn: Connection, targetConn: Connection, tableName: String): Int {
        val targetColumns = getTableColumns(targetConn, tableName)
        val targetColumnSet = LinkedHashSet(targetColumns)
        val sourceColumns = getTableColumns(sourceConn, tableName)
        val commonColumns = sourceColumns.filter { targetColumnSet.contains(it) }
        if (commonColumns.isEmpty()) {
            throw SQLException("No common columns found for table $tableName")
        }
        val columnList = commonColumns.joinToString(", ")
        val placeholders = commonColumns.joinToString(", ") { "?" }
        val selectSql = SQL_SELECT + columnList + SQL_FROM + tableName
        val insertSql = "INSERT OR REPLACE INTO $tableName ($columnList$SQL_VALUES$placeholders)"

        sourceConn.createStatement().use { sourceStmt ->
            sourceStmt.executeQuery(selectSql).use { sourceResult ->
                targetConn.prepareStatement(insertSql).use { targetStatement ->
                    val columnCount = commonColumns.size
                    var syncedRows = 0
                    while (sourceResult.next()) {
                        for (i in 1..columnCount) {
                            targetStatement.setObject(i, sourceResult.getObject(i))
                        }
                        targetStatement.addBatch()
                        syncedRows++
                    }
                    targetStatement.executeBatch()
                    return syncedRows
                }
            }
        }
    }

    @Throws(SQLException::class)
    private fun replaceTable(sourceConn: Connection, targetConn: Connection, tableName: String): Int {
        val targetColumns = getTableColumns(targetConn, tableName)
        val targetColumnSet = LinkedHashSet(targetColumns)
        val sourceColumns = getTableColumns(sourceConn, tableName)
        val commonColumns = sourceColumns.filter { targetColumnSet.contains(it) }
        if (commonColumns.isEmpty()) {
            throw SQLException("No common columns found for table $tableName")
        }

        val columnList = commonColumns.joinToString(", ")
        val placeholders = commonColumns.joinToString(", ") { "?" }
        val selectSql = SQL_SELECT + columnList + SQL_FROM + tableName
        val deleteSql = "DELETE FROM $tableName"
        val insertSql = "$SQL_INSERT_INTO$tableName ($columnList$SQL_VALUES$placeholders)"

        val originalAutoCommit = targetConn.autoCommit
        try {
            sourceConn.createStatement().use { sourceStmt ->
                sourceStmt.executeQuery(selectSql).use { sourceResult ->
                    targetConn.createStatement().use { deleteStatement ->
                        targetConn.prepareStatement(insertSql).use { targetStatement ->
                            targetConn.autoCommit = false
                            deleteStatement.executeUpdate(deleteSql)
                            val columnCount = commonColumns.size
                            var syncedRows = 0
                            while (sourceResult.next()) {
                                for (i in 1..columnCount) {
                                    targetStatement.setObject(i, sourceResult.getObject(i))
                                }
                                targetStatement.addBatch()
                                syncedRows++
                            }
                            targetStatement.executeBatch()
                            targetConn.commit()
                            return syncedRows
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            targetConn.rollback()
            throw e
        } finally {
            targetConn.autoCommit = originalAutoCommit
        }
    }

    @Throws(SQLException::class)
    private fun syncConfiguration(sourceConn: Connection, targetConn: Connection, includeExternalPlayerPaths: Boolean): Boolean {
        val sourceColumns = getTableColumns(sourceConn, DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName)
        val targetColumns = getTableColumns(targetConn, DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName)
        val targetColumnSet = LinkedHashSet(targetColumns)
        val commonColumns = sourceColumns.filter { targetColumnSet.contains(it) && !CONFIGURATION_SYNC_EXCLUDED_COLUMNS.contains(it) }
        if (commonColumns.isEmpty()) {
            return false
        }

        val sourceRow = readFirstConfigurationRow(sourceConn, commonColumns) ?: return false
        val targetRow = readFirstConfigurationRow(targetConn, commonColumns)
        if (!includeExternalPlayerPaths && targetRow != null) {
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath1")
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath2")
            sourceRow.copyColumnIfPresentFrom(targetRow, "playerPath3")
            sourceRow.copyColumnIfPresentFrom(targetRow, "defaultPlayerPath")
        }
        upsertFirstConfigurationRow(targetConn, sourceRow, targetRow?.id)
        return true
    }

    @Throws(SQLException::class)
    private fun syncPublishedM3uSelections(sourceConn: Connection, targetConn: Connection): Int =
        syncSelectionTableWithAccountRemap(sourceConn, targetConn, DatabaseUtils.DbTable.PUBLISHED_M3U_SELECTION_TABLE.tableName, emptyList())

    @Throws(SQLException::class)
    private fun syncSelectionTableWithAccountRemap(
        sourceConn: Connection,
        targetConn: Connection,
        selectionTable: String,
        extraColumns: List<String>
    ): Int {
        val accountTable = DatabaseUtils.DbTable.ACCOUNT_TABLE.tableName
        val originalAutoCommit = targetConn.autoCommit
        var syncedRows = 0
        val selectionColumns = buildSelectionColumnList(extraColumns)
        val placeholders = buildSelectionPlaceholders(extraColumns.size + 1)

        try {
            sourceConn.createStatement().use { sourceSelections ->
                sourceSelections.executeQuery(SQL_SELECT + selectionColumns + SQL_FROM + selectionTable + " ORDER BY id").use { sourceRows ->
                    sourceConn.prepareStatement("SELECT accountName FROM $accountTable WHERE id = ?").use { sourceAccountName ->
                        targetConn.prepareStatement("SELECT id FROM $accountTable WHERE accountName = ?").use { targetAccountId ->
                            targetConn.createStatement().use { deleteTargetSelections ->
                                targetConn.prepareStatement("$SQL_INSERT_INTO$selectionTable ($selectionColumns$SQL_VALUES$placeholders)").use { insertTargetSelection ->
                                    targetConn.autoCommit = false
                                    deleteTargetSelections.executeUpdate("DELETE FROM $selectionTable")
                                    while (sourceRows.next()) {
                                        val targetSelectionAccountId = resolveTargetAccountId(sourceRows, sourceAccountName, targetAccountId)
                                        if (targetSelectionAccountId != null) {
                                            insertTargetSelection.setString(1, targetSelectionAccountId)
                                            for (i in extraColumns.indices) {
                                                insertTargetSelection.setObject(i + 2, sourceRows.getObject(extraColumns[i]))
                                            }
                                            insertTargetSelection.addBatch()
                                            syncedRows++
                                        }
                                    }
                                    if (syncedRows > 0) {
                                        insertTargetSelection.executeBatch()
                                    }
                                    targetConn.commit()
                                    return syncedRows
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: SQLException) {
            targetConn.rollback()
            throw e
        } finally {
            targetConn.autoCommit = originalAutoCommit
        }
    }

    private fun buildSelectionColumnList(extraColumns: List<String>): String {
        val columns = ArrayList<String>()
        columns.add(ACCOUNT_ID)
        columns.addAll(extraColumns)
        return columns.joinToString(", ")
    }

    private fun buildSelectionPlaceholders(count: Int): String =
        (0 until count).joinToString(", ") { "?" }

    @Throws(SQLException::class)
    private fun readFirstConfigurationRow(conn: Connection, columns: List<String>): ConfigurationRow? {
        val columnList = columns.filter { !it.equals("id", true) }.joinToString(", ")
        if (columnList.isBlank()) {
            return null
        }
        val sql = SQL_SELECT + "id, " + columnList + SQL_FROM + DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName + " ORDER BY id LIMIT 1"
        conn.createStatement().use { stmt ->
            stmt.executeQuery(sql).use { rs ->
                if (!rs.next()) {
                    return null
                }
                val row = ConfigurationRow(rs.getObject("id"))
                for (column in columns) {
                    if (column.equals("id", true)) continue
                    row.put(column, rs.getObject(column))
                }
                return row
            }
        }
    }

    @Throws(SQLException::class)
    private fun upsertFirstConfigurationRow(conn: Connection, row: ConfigurationRow, targetId: Any?) {
        val columns = ArrayList(row.columns())
        if (columns.isEmpty()) {
            return
        }
        if (targetId == null) {
            val columnList = columns.joinToString(", ")
            val placeholders = columns.joinToString(", ") { "?" }
            val sql = SQL_INSERT_INTO + DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName + " (" + columnList + SQL_VALUES + placeholders + ")"
            conn.prepareStatement(sql).use { statement ->
                bindColumns(statement, columns, row)
                statement.executeUpdate()
            }
            return
        }

        val assignments = columns.joinToString(", ") { "$it=?" }
        val sql = "UPDATE " + DatabaseUtils.DbTable.CONFIGURATION_TABLE.tableName + " SET " + assignments + " WHERE id = ?"
        conn.prepareStatement(sql).use { statement ->
            bindColumns(statement, columns, row)
            statement.setObject(columns.size + 1, targetId)
            statement.executeUpdate()
        }
    }

    @Throws(SQLException::class)
    private fun bindColumns(statement: PreparedStatement, columns: List<String>, row: ConfigurationRow) {
        for (i in columns.indices) {
            statement.setObject(i + 1, row.get(columns[i]))
        }
    }

    @Throws(SQLException::class)
    private fun resolveTargetAccountId(
        sourceRows: ResultSet,
        sourceAccountName: PreparedStatement,
        targetAccountId: PreparedStatement
    ): String? {
        val sourceAccountId = sourceRows.getString(ACCOUNT_ID)
        if (sourceAccountId.isNullOrBlank()) {
            return null
        }
        sourceAccountName.setString(1, sourceAccountId)
        sourceAccountName.executeQuery().use { sourceAccount ->
            if (!sourceAccount.next()) {
                return null
            }
            val accountName = sourceAccount.getString("accountName")
            if (accountName.isNullOrBlank()) {
                return null
            }
            targetAccountId.setString(1, accountName)
            targetAccountId.executeQuery().use { targetAccount ->
                return if (targetAccount.next()) targetAccount.getString("id") else null
            }
        }
    }

    @Throws(SQLException::class)
    private fun getTableColumns(conn: Connection, tableName: String): List<String> {
        val columns = ArrayList<String>()
        conn.createStatement().use { stmt ->
            stmt.executeQuery("PRAGMA table_info($tableName)").use { rs ->
                while (rs.next()) {
                    columns.add(rs.getString("name"))
                }
            }
        }
        return columns
    }

    private class ConfigurationRow(val id: Any?) {
        private val values = LinkedHashMap<String, Any?>()
        fun put(column: String, value: Any?) {
            values[column] = value
        }
        fun get(column: String): Any? = values[column]
        fun columns(): Set<String> = values.keys
        fun copyColumnIfPresentFrom(source: ConfigurationRow?, column: String) {
            if (source != null && values.containsKey(column) && source.values.containsKey(column)) {
                values[column] = source.values[column]
            }
        }
    }
}
